libsolidity/analysis/ContractLevelChecker.cpp

```diff
@@ -135,50 +214,58 @@ void ContractLevelChecker::findDuplicateDefinitions(map<string, vector<T>> const
 
 void ContractLevelChecker::checkIllegalOverrides(ContractDefinition const& _contract)
 {
-	// TODO unify this at a later point. for this we need to put the constness and the access specifier
-	// into the types
-	map<string, vector<FunctionDefinition const*>> functions;
-	map<string, ModifierDefinition const*> modifiers;
+	FunctionSet const& funcSet = getBaseFunctions(&_contract);
+	ModifierSet const& modSet = getBaseModifiers(&_contract);
 
-	// We search from derived to base, so the stored item causes the error.
-	for (ContractDefinition const* contract: _contract.annotation().linearizedBaseContracts)
+	checkModifierOverrides(funcSet, modSet, _contract.functionModifiers());
+
+	for (FunctionDefinition const* function: _contract.definedFunctions())
 	{
-		for (FunctionDefinition const* function: contract->definedFunctions())
-		{
-			if (function->isConstructor())
-				continue; // constructors can neither be overridden nor override anything
-			string const& name = function->name();
-			if (modifiers.count(name))
-				m_errorReporter.typeError(modifiers[name]->location(), "Override changes function to modifier.");
+		if (contains_if(modSet, MatchByName{function->name()}))
+			m_errorReporter.typeError(function->location(), "Override changes modifier to function.");
 
-			for (FunctionDefinition const* overriding: functions[name])
-				checkFunctionOverride(*overriding, *function);
+		// Skip if not overridable
+		if (!function->isOverridable())
+			continue;
 
-			functions[name].push_back(function);
-		}
-		for (ModifierDefinition const* modifier: contract->functionModifiers())
+		// No overriding functions found
+		if (funcSet.find(function) == funcSet.cend())
 		{
-			string const& name = modifier->name();
-			ModifierDefinition const*& override = modifiers[name];
-			if (!override)
-				override = modifier;
-			else if (ModifierType(*override) != ModifierType(*modifier))
-				m_errorReporter.typeError(override->location(), "Override changes modifier signature.");
-			if (!functions[name].empty())
-				m_errorReporter.typeError(override->location(), "Override changes modifier to function.");
+			// Skip if no override
+			if (!function->overrides())
+				continue;
```

Comment:
Do we need this complex logic? Would `checkOverrideList` also work if we would not have  `continue` here?

Reply:
This is partly a simplification because in all continue cases we try to access `function->overrides()` and we don't want to check for nullptr for every access

Reply:
hmm no that's not correct, we do check for nullptr in `checkOverrideList`. However, the output of the error message just a few lines lower needs that check and we dereference function after this if-body, if we continue we have to check differently that `function` is not nullptr..

I feel avoiding this `continue` would create too many other cases that need to be checked

Reply:
What about putting an `else` in front of the `checkOverrideList` call?

As far as I see it, the whole code inside the `if (funcSet.find(function) == funcSet.cend())` is just there to have a better error message, right? `checkOverrideList` would also determine that the list needs to be empty but is not and report an error in that case. Maybe we could just move the improved error inside the `checkOverrideList` function and remove this special case here?